design/ceph/object/ceph-bucket-notification-crd.md

```diff
@@ -0,0 +1,62 @@
+# Ceph bucket notifications CRD
+
+## Overview
+
+In its new release, Ceph introduced the bucket notifications feature. It allows to send messages to various endpoints when a new event occurs on a bucket (ref: <https://docs.ceph.com/docs/master/radosgw/notifications/)>
+
+Setup of those notifications are normally done by sending HTTP requests to the RGW, either to create/delete topics pointing to specific endpoints, or create/delete bucket notifications based on those topics.
+
+This functionality eases this process by avoiding to use external tools or scripts. It is replaced by creation of CR definitions that  contain all the information necessary to create topics and/or notifications, which the rook operator processes.
+
+## Goals
+
+Creates a CRD for topics and a CRD for notifications, defining all the necessary and optional information for the various endpoints.
+
+Extends the rook operator to handle the CRs that would be submitted by users.
+
+## Implementation
+
+The CR for a topic configuration takes this form:
+
+```yaml
+apiVersion: ceph.rook.io/v1
+kind: Topic
+metadata:
+  name: # name of the topic
+  namespace: # namespace where topic belongs
+Spec:
+  Endpoint: #(mandatory) URI of an endpoint to send push notification to
+  Port: #(optional) port used to communicate, if empty uses the default ones
+  http:
+    verify-ssl: #indicate whether the server certificate is validated by the client or not (`true` by default)
+  amqp:
+    vhost: # defaults to `/`
+    ackLevel: # none/routable/broker, optional (default - broker)
+    amqp-exchange: # exchanges must exist and be able to route messages based on topics
+  kafka:
+    use-ssl: # secure connection will be used for connecting with the broker (`false` by default)
+    ca-location: # this specified CA will be used, instead of the default one, to authenticate the broker
+    ackLevel: # none/broker, optional (default - broker)
+```
+
+The CR for bucket notification takes this form:
+
+```yaml
+apiVersion: ceph.rook.io/v1
+kind: BucketNotification
+metadata:
+  name: # name of the notification
+  namespace: # namespace where notification belongs
+Spec:
+  Topic: #(mandatory) reference to the topic, topic_arn
+  Filter: #(optional) Prefix/Suffix/Regex/Metadata/Tags, optional (default - {})
+    Metadata:
+    - name:
+      value:
+    Prefix:
+    - name:
+      value:
+  Events: # vaules appilicable [here](https://docs.ceph.com/en/latest/radosgw/s3-notification-compatibility/#event-types), (default all)
+  bucketName: # name(s) of bucket in which this notification send for
```

Comment:
why not keeping the definition here generic, and referencing from the bucket definition?

Reply:
My plan is to keep `bucketnofication` CR in the same namespace where rgw pod exists or tie it to rook-ceph-objectstore service than a `OBC` or `BR` in upcoming COSI design . In the same namespace there can be multiple rgw pods running, thats why I have given option to specify `endpoint`

Reply:
if you tie the definition to a specific OBC, then add it there.
you won't need `bucketName` and `endpoint` here